### DISCUSSION TITLE
High Memory Usage by next-server process.

### DISCUSSION THREAD

**Discussion by ShivamArora** [vercel/next.js#70178] [Help]
### Summary

We're experiencing significantly high memory usage by the `next-server` process when using Turbopack locally.

Initially when the server is started using `next dev --turbo`, the process is well under 250MB. But as soon as we open a page and the compilation starts for the page, the memory usage jumps to ~6.5GB and then doesn't come down even after the page has been compiled and displayed and the system is in an idle state.

On making changes to the codebase with HMR, we've seen the memory usage jumping to even 10-11GB at times which makes it difficult to work with the project as the system becomes slow.

Comparatively when we don't use turbopack and start the project with Webpack, it takes around 3.5GB-4GB when a page is loaded and turbopack is using comparatively more than 1.5x memory than Webpack.

We haven't been able to identify what is the root cause behind such high memory consumption. And it is making tremendously difficult to use turbopack with our project.

### Additional information

```js
`next info` output:


Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:13:04 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6020
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 18.20.4
  npm: 10.7.0
Relevant Packages:
  next: 14.2.11 // Latest available version is detected (14.2.11).
  eslint-config-next: 14.2.11
  react: 18.3.1
  react-dom: 18.3.1
Next.js Config:
  output: N/A
```
```


### Example

_No response_

**Comment by raju-sirigineedi** [comment]
Hey @ShivamArora We are experiencing the same problem. Did you find any workaround for it. ?

**Reply by ShivamArora** [reply]
Hey @raju-sirigineedi , haven't found any workarounds for it. Found 2-3 issues in the next forums from the discord server, but seems like people are asking to accept the high memory usage instead of fixing it.

Please let us know in case you find a fix for this. As a temporary measure, we're sticking to webpack as it is taking lesser memory than turbopack contrary to the claims by turbopack.

**Reply by raju-sirigineedi** [reply]
Hi @ShivamArora, try upgrading to next 15. I mean the migration might be rough since our codebases are too big, but its worth the shot. We are feeling good dev experience compared to 14. 
We created an experimental branch to our repo with next 15, and followed their guide to migrate. Its good.

**Reply by ShivamArora** [reply]
I'm looking forward to the same @raju-sirigineedi . Just blocked on storybook as it hasn't completely rolled out support for Next 15 at the moment. Will try upgrading once they roll out support and share feedback.

**Comment by teknology** [comment]
I'm having the same issue. It's as if Next has a memory leak.

**Comment by pnogier** [comment]
Hi there, it's been a long time that I have the same issue. 

We have a pretty big Next project, but it is impossible to work with in dev mode. It goes up to more than 13gb of memory usage at each fast refresh. Even tho my mac (M1 Pro) has 16gb unified memory, so it goes up to about 20gb in swap. 

For real this is getting crazy. I am used to have slowness issues as I often work with React Native on iOS and Android simulators, running XCode and Android Studio at the same time, but wow, it is so fast compared to just using Next in dev mode... 

Next seems to not be scalable at all, or have a memory leak somewhere, maybe it is on our side tho.. 
We had this issue on every previous version since next `12.x.x`. Now we are using `15.3.2`.

It would be really helpful if someone has a workaround, or even a bit of an idea of what could be happening.

Thanks in advance,
Have a great day

**Comment by jelling** [comment]
Same issue here.  Most promising issue tracking this seems to be: https://github.com/vercel/next.js/issues/78069

**Comment by brillaintlcd** [comment]
This issue seems to be fixed in the newly released version [v16.3](https://nextjs.org/blog/next-16-3). I'm afraid, I haven't found any other solutions for the older versions.

This solution worked for me and it dropped my average memory consumption from (~8-10GB) to (~1.8-2GB). Unlike before, where it used to never cleanup memory and constantly consume more and more, it now also actively optimizes memory usage - which is amazing as well.

**Comment by pnogier** [comment]
@brillaintlcd Indeed, fixed and wow what a relief. At a point next server was taking up to 45gb of ram, and making my mac unusable. 
`v16.3` is impressive as I never exceed 1.5/2 gb now. 
Thanks a lot to everyone who worked on this, even if it took time..
